### ISSUE TITLE
Step state persistence Issue 

### ISSUE THREAD

**Issue by SidIcarus** [Codeception/Codeception#4061]
#### What are you trying to achieve?
I have a Cest-format sign in error message tests running using data providers. I'm unsure why the tests fail at the moment but I am certain it is related with our logic of the signing and not Codeception.  I have tested this using `var_dump` in my test to see what the message I actually have is as well as by switching the position of my examples in the data provider, as well a adding in another example (a third in my case). The error persists.

#### What do you get instead?
Instead of the proper dataprovider given example data, it will show the last array of example data on all tests that have failed in the step section of the failure info. 

> Image of my console output
<img width="784" alt="codeception_error" src="https://cloud.githubusercontent.com/assets/3842798/23973332/0b74ff14-09ac-11e7-9598-2ccdc926a3ab.png">

* Codeception version: 2.2.9
* PHP Version: 7.1
* Operating System: macOS Sierra 10.12.1
* Installation type: Composer

**Comment by edno** [comment]
Looks like a bug. Can you provide an example of code?

**Comment by SidIcarus** [comment]
@edno 

```php
<?php

namespace Main\Tests;

use Main\PhpBrowser as Tester;
use Main\Page\PhpBrowser\Signin as Page;
use Codeception\Example as Example;

class SigninCest {
    protected $page;

    // Using _inject to set the page doesn't work
    public function _before(Tester $I, Page $page) {
        $this->page = $page;
    }

    protected function passwordProvider() {
        return [
            ['passwordError_Min', '1234567', '12345678'],
            ['passwordError_Empty', ' ', '12345678']
        ];
    }

    protected function seeError(Tester $I, string $labelID, string $msgID,
        bool $see = true) {
        $label = $this->page::$ui[$labelID];
        $msg = $this->page::$msgs[$msgID];

        $this->page->clickOnSignin();
        var_dump($msg);
        $see ? $I->see($msg) : $I->dontSee($msg);
    }
    
    /**
     * @dataprovider passwordProvider
     */
    public function seePasswordError(Tester $I, Example $ex) {
        $this->page->enterPassword($ex[1]);
        $this->seeError($I, 'passwordErrorLabel', $ex[0]);
    }
}
```

**Comment by edno** [comment]
I've made a quick test with a code similar to yours and I don't reproduce the issue. 

I can see that you are calling a class property using static notation `::$` instead of `->`, so I would suspect that the persistence has something to do with the code of your class **Main\Page\PhpBrowser\Signin**.

**Comment by SidIcarus** [comment]
I do not believe that is the case.
Here is my code, it is intended for the reusability of my page object between all test suites.

```php
namespace Main\Page;
abstract class Page {
    public static $title, $url, $ui;
    protected $tester;
    // Late static binding allows for the setting of inherited static variables 
    // to resolve to the subclass only.
    protected function __construct($title, $url, $ui) {
        static::$title = $title;
        static::$url = $url;
        static::$ui = $ui;
    }
    public static function route($param) { return static::$url.$param; }
}

namespace Main\Page\PhpBrowser;
class Signin extends \Main\Page\Page {
    use \Main\Page\Base\Signin;
    public function __construct(\Main\PhpBrowserTester $tester) {
        $this->tester = $tester;
        list ($title, $url, $ui) = $this->init();
        parent::__construct($title, $url, $ui);
    }
}
```

```php
namespace Main\Page;
trait PageTrait { abstract protected function init(); }

namespace Main\Page\Base;
trait Signin { use \Main\Page\PageTrait;
    public function init() {
        $ui = array ( 'array' => 'ofStuff');
        return array ('PageTitle', 'PageUrl', $ui);
    }
```

Regardless of the above, you can see at the top of my screenshot the underlined output of $msg which indicates that the dataprovider is passing the right example during runtime.

**Comment by edno** [comment]
Static inheritance is not the way to do it. I don't understand what you try to achieve by using static in your example:
```php
   public static $title, $url, $ui;
    protected $tester;
    // Late static binding allows for the setting of inherited static variables
    // to resolve to the subclass only.
    protected function __construct($title, $url, $ui) {
        static::$title = $title;
        static::$url = $url;
        static::$ui = $ui;
    }
```
The purpose of using `static` is:
> Declaring class properties or methods as static makes them accessible without needing an instantiation of the class.

Or in your example you instantiate your class. My suggestion would be to do:
```php
    public $title, $url, $ui; // you can also make them protected and provide get() and set()
    protected $tester;
    // Late static binding allows for the setting of inherited static variables
    // to resolve to the subclass only.
    protected function __construct($title, $url, $ui) {
        $this->title = $title;
        $this->url = $url;
        $this->ui = $ui;
    }
```
There's a good discussion about static in PHP on [Stackoverflow](http://stackoverflow.com/questions/532754/inheritance-of-static-members-in-php).

**Comment by SidIcarus** [comment]
I don't think you realize that my code is an abstraction of the example PageObject [PageObject docs](http://codeception.com/docs/06-ReusingTestCode#PageObjects).

**Comment by edno** [comment]
Static has nothing to do with abstraction. In documentation example, you can clearly see that the static parameters are indeed static, hence these are not initialised later or updated later.

[Here](https://github.com/edno/kleis/blob/develop/tests/_support/Page/KleisPage.php) you can see another example of use of the page object pattern without static.

**Comment by edno** [comment]
> Here is a discussion about the purpose of late static binding on Stackoverflow as well as in the php manual

Late static binding is used for overriding static by other static. In your example, you expect static to behave like a standard parameter (using late initialisation and update), that is not the purpose of static and I would bet that the reason of your issue.

**Comment by SidIcarus** [comment]
In other parts of my code I access those static properties as intended. 

Regardless, I generated an empty cept test in my other test suite. It contains nothing but what is there when generated with `g:cept`.  As you can see, the step it fails on shows `See "Password is Required."` which is obviously not the case.

<img width="505" alt="test in different suite" src="https://cloud.githubusercontent.com/assets/3842798/24019125/267c7186-0a6d-11e7-80a2-f653ea0b075d.png">

**Comment by SidIcarus** [comment]
@edno 
I have also converted all of my code to use instance variables instead of static variables and the issue still persists.

> I would bet that is the reason of your issue.

Fortunately, you did not bet.

**Comment by ThomasLandauer** [comment]
I can confirm the initial issue, I was just about to report it myself.

I run Codeception 2.3.5 on PHP 7.1.

In my case, I neither have a data provider, nor anything "fancy" like abstract/static/trait/whatever - just some good old public functions :-)

I'm suspecting the error is in Codeception's reporting system: The *last* failed test seems to overwrite the `Step` info of all previously failed tests.

My output shows this even clearer than SidIcarus':
```
There were 3 failures:
---------
1) MeineDatenFunctionalCest: Try ohne token
 Test  tests\functional\MeineDatenFunctionalCest.php:tryOhneToken
 Step  Don't see element {"css":".fehlermeldungFalscherToken"}
 Fail  Expected HTTP Status Code: 401 (Unauthorized). Actual Status Code: 200 (OK)
Failed asserting that 200 matches expected 401.
---------
2) MeineDatenFunctionalCest: Try richtiger token und logout
 Test  tests\functional\MeineDatenFunctionalCest.php:tryRichtigerTokenUndLogout
 Step  Don't see element {"css":".fehlermeldungFalscherToken"}
 Fail  Failed asserting that '/meine-kuendigungen' contains "/meine-kuendigung/1".
---------
3) MeineDatenFunctionalCest: Try keine kuendigungen vorhanden
 Test  tests\functional\MeineDatenFunctionalCest.php:tryKeineKuendigungenVorhanden
 Step  Don't see element {"css":".fehlermeldungFalscherToken"}
 Fail  Element '"css":".fehlermeldungFalscherToken"' was found
```

If you compare the `Step` line with the `Fail` line, you can easily see that they're talking about completely different things (in test (1) and (2)). Only in test (3) it makes sense.

**Comment by Naktibalda** [comment]
The issue as described by @ThomasLandauer was fixed in November 2017 and released in Codeception 2.3.7

See #4654
